### PR TITLE
Upstream sync 21/N: merge 50ba4bc6b2 [Feature] Mask Replay (#49577) (conflict)

### PR DESCRIPTION
**Pull request by roberteg16**
## Context

Twenty-first step of the batched upstream catch-up. Stacked on #<batch 20 PR>.

**Conflict-only** step.

Merged upstream changes:
1. `50ba4bc6b2` #49577 — [Feature] Mask Replay

One conflict, in `vllm/config/model.py`, inside the `ignored_factors` set used to build the model config's cache key — two independent one-line additions:

```
ours:     "w4a16_prefill_dequant",
theirs:   "return_sampling_mask",
resolved: "w4a16_prefill_dequant",     <- ours
          "return_sampling_mask",      <- theirs
```

Union.

## Audit

Verified both names refer to real fields on the same class before keeping them: `w4a16_prefill_dequant` at `model.py:249` (the fork's W4A16 prefill dequant mode) and `return_sampling_mask` at `model.py:255` (new in this commit).

The direction of risk is worth stating because it is the opposite of the usual import case: a *stale* entry in this set would be silently harmless, but a **missing** one is not. The set lists fields deliberately excluded from the cache key, so dropping either side folds a non-semantic field into the hash and causes spurious recompiles whenever it changes — a slow, invisible regression rather than a crash.

`py_compile` and `ruff check` pass.

**Merge commit only — do not squash or rebase.**

AI assistance was used to prepare this merge.

## Test plan

- [x] Both names verified to be real fields (`model.py:249`, `model.py:255`)
- [x] `py_compile` + `ruff check`
- [x] Build + 6-model benchmark sweep vs the Strix Halo dashboard — no regression on transformers 5.15.0 ([results](https://github.com/ROCm/vllm/pull/1268#issuecomment-5581543544))

### CODE REVIEW

**Comment by roberteg16** [comment]
## Six-model verification sweep — no regression

Run against the CI wheel built from this branch, with **transformers 5.15.0**
(the version this stack unpins, in batch 11 / #1258).

| item | value |
|---|---|
| wheel | `vllm-0.26.1rc1.dev1938+gee4949875` ([run 34105303446](https://github.com/ROCm/vllm/actions/runs/34105303446), branch `rogarcia.merge-upstream-117`) |
| transformers | **5.15.0** |
| torch / torchvision / torchaudio | `2.13.0` / `0.28.0` / `2.11.0`, all `+rocm10.1.0a20260906` |
| triton | `3.8.0+git4cff872c.rocm10.1.0a20260906` |
| GPU | AMD Radeon 8060S (gfx1151) |

Everything except `transformers` matches the Strix Halo reference run
`regression_20260907_093058`; baselines below are that run's numbers, i.e. the
**same board**, not the worked example in the skill docs.

| bench | test id | decode tok/s | baseline | Δ |
|---|---|---:|---:|---:|
| bench1 | `Qwen3-30B-A3B-Instruct-2507-AWQ-4bit_128` | 78.8 | 80.2 | −1.7% ¹ |
| bench2 | `Qwen3.6-35B-A3B-W4A16-llmcompressor_VLM_dyn-lm-int8g32` | 83.2 | 82.9 | +0.4% |
| bench3 | `Gemma-3-4B-IT_VLM_w4a16` | 61.4 | 61.6 | −0.3% |
| bench4 | `Qwen3-Omni-30B-A3B-Instruct_VLM_AWQ-4bit` | 76.4 | 75.5 | +1.2% |
| bench5 | `Qwen2.5-0.5B-Instruct_128` | 359.3 | 357.0 | +0.6% |
| bench6 | `Gemma-4-31B-IT_VLM_AWQ-4bit` | 10.3 | 10.3 | +0.0% |

All six: 10/10 successful requests, 0 failures.

¹ **bench1 is measurement noise, not a regression.** Four runs of the same
configuration: 79.5 / 78.8 / 79.9 / 78.7 — a 1.2 tok/s (1.5%) spread *within*
one configuration, wider than the gap to baseline, and its best run is −0.4%.
Runs with and without the fix below overlap, so the two cannot be separated at
this benchmark's resolution. Worth noting for future sweeps: bench1's own
variance exceeds the 1% decode gate, so that gate is not meaningful for it
as-is.

### A real regression was found and fixed in this stack

The first sweep showed **bench2 at 80.2 tok/s (−3.3%)**, reproduced four times
(80.2 / 80.2 / 80.3 / 80.2 — spread 0.1%). Bisected with the per-PR CI wheels:

| wheel | batch | decode |
|---|---|---:|
| `-107` (#1258) | 11 | 83.1 |
| `-110` (#1261) | 14 | 80.3 |
| `-112` (#1263) | 16 | 80.3 |
| `-117` (#1268) | 21 | 80.2 |

Attributed to **batch 13** (upstream `1ab2801dde`, "[ROCm] Remove stale SDPA and
skinny GEMM workarounds"), which adds `weight.is_contiguous()` to the
`use_skinny` gate. That disqualifies every weight
`pad_weights_avoid_cache_cliff_on_gfx11` has padded — the pad changes `stride(0)`
only, deliberately, to dodge the gfx11x L2/MALL channel-hash cliff — and drops
those layers onto the fallback GEMM. A runtime trace showed **1760 of 4150 calls**
being rejected this way.

Fixed in `f94abd3f06` on the batch-13 branch and propagated up the stack:

```
skinny_operands_compatible = weight.stride(-1) == 1 and (
    bias is None or bias.is_contiguous()
)
```

The kernel needs K contiguous, not the whole tensor: the host captures
`Kap/Kbp = stride(0)` (`skinny_gemms.cu:1264`) and the kernel indexes
`B_[row * Kbp]` and `s[k_ + Kap * n]` rather than assuming `K`; the LDS-fit
check in the launcher is sized with `Kbp` as well.

**Correctness was verified, not assumed** — if the kernel were wrong on padded
strides then the pre-batch-13 numbers were wrong too. `ops.wvSplitK` with a
padded weight vs `torch.nn.functional.linear`, four shapes:
`1x2560x2048`, `1x4096x4096`, `2x2560x2048`, `4x5120x2048` → max |Δ| **0.0039**,
i.e. fp16 accumulation noise. The runtime trace after the fix shows all 1760
previously-rejected calls have `stride(-1) == 1` and are now eligible.

AI assistance was used to prepare this verification.

**Review by eble-amd — APPROVED** [review]
